### PR TITLE
Inference with AD3 on CRF models now supports hard logic constraints.

### PR DESCRIPTION
**Pull request by jlmeunier**
I wanted to exploit structural (logical) constraints at inference time on a EdgeFeatureGraphCRF model trained with a OneSlackSSVM learner.

So I extended AD3 as explained in their ICML 2011 paper, and in turn I extended pystruct to have an optional _constraints_ parameter in the _predict_ method.

I did that extension on the CRF and SSVM classes. So almost it is widely useable.
But this inference under constraints is only supported with AD3.
(Note: This code requires the updated AD3 library, currently at https://github.com/jlmeunier/AD3. 
A pull request has been done as well.)

The auto-test produces the same results, and in my application works fine.

JL

### CODE REVIEW

**Comment by jlmeunier** [comment]
Hi again,

A new update: this time I'm contributing a new model where nodes can be of different nature (and by consequence edges as well). Each node type has its own separate state-space and weight vector (edge types also have separate weight vectors).

This required updates to AD3, which I did as well.

Thanks

JL

**Comment by amueller** [comment]
hey. Sorry I'm swamped. I'll try to review but I can't promise anything. It would be good to have the tests work, though I'm sure it's my fault and not yours.

File: examples/logs/plot_snakes.log [hunk 1]
```diff
@@ -0,0 +1,27 @@
+Please be patient. Learning will take 5-20 minutes.
```
Comment (amueller) [inline examples/logs/plot_snakes.log]: is there a reason you committed the logs?

File: setup.py [hunk 2]
```diff
@@ -10,18 +10,18 @@
 include_dirs = [np.get_include()]
 
 setup(name="pystruct",
-      version="0.2.5",
-      install_requires=["ad3"],
+      version="0.3.3",
```
Comment (amueller) [inline setup.py]: what is that based on?

File: pystruct/models/typed_crf.py [hunk 3]
```diff
@@ -0,0 +1,299 @@
...
+    it under the terms of the GNU General Public License as published by
+    the Free Software Foundation, either version 3 of the License, or
+    (at your option) any later version.
+
+    This program is distributed in the hope that it will be useful,
+    but WITHOUT ANY WARRANTY; without even the implied warranty of
+    MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+    GNU General Public License for more details.
+
+    You should have received a copy of the GNU General Public License
```
Comment (amueller) [inline pystruct/models/typed_crf.py]: Please relicense it under BSD or MIT.

**Comment by jlmeunier** [comment]
Hi Andreas

Not sure I get properly your question.

I went to 0.3 because of the add of new functionalities
0.3.3 because of 3rd big stable SW state

File: examples/logs/plot_snakes.log (hunk 1 above)
Comment (jlmeunier) [inline examples/logs/plot_snakes.log]: simply to preserve them as a snapshot of current behavior. But we can get rid of them.

File: pystruct/models/typed_crf.py (hunk 3 above)
Comment (jlmeunier) [inline pystruct/models/typed_crf.py]: ok, will do, fine with me

**Comment by jlmeunier** [comment]
Hi Andy

Thanks for the review.

I pushed 0.3.4 which fixes ones pb and now passes all tests (apart one which was not passing anyway, and fixing it would require quite some work for me to understand it fully)

JL

**Pull request closed without merging**